client/alloc_endpoint.go

```diff
@@ -100,3 +116,186 @@ func (a *Allocations) Stats(args *cstructs.AllocStatsRequest, reply *cstructs.Al
+	defer conn.Close()
+
+	// Decode the arguments
+	var req cstructs.AllocExecRequest
+	decoder := codec.NewDecoder(conn, structs.MsgpackHandle)
+	encoder := codec.NewEncoder(conn, structs.MsgpackHandle)
+
+	if err := decoder.Decode(&req); err != nil {
+		handleStreamResultError(err, helper.Int64ToPtr(500), encoder)
+		return
+	}
+
+	aclObj, token, err := a.c.resolveTokenAndACL(req.QueryOptions.AuthToken)
+	{
+		// log access
+		execID := uuid.Generate()
+		tokenName, tokenID := "", ""
+		if token != nil {
+			tokenName, tokenID = token.Name, token.AccessorID
+		}
+
+		a.c.logger.Info("task exec session starting",
+			"exec_id", execID,
+			"alloc_id", req.AllocID,
+			"task", req.Task,
+			"command", req.Cmd,
+			"tty", req.Tty,
+			"access_token_name", tokenName,
+			"access_token_id", tokenID,
+		)
+
+		defer a.c.logger.Info("task exec session ended", "exec_id", execID)
+	}
+
+	// Check read permissions
+	if err != nil {
+		handleStreamResultError(err, nil, encoder)
+		return
+	} else if aclObj != nil {
+		exec := aclObj.AllowNsOp(req.QueryOptions.Namespace, acl.NamespaceCapabilityAllocExec)
+		if !exec {
+			handleStreamResultError(structs.ErrPermissionDenied, nil, encoder)
+			return
+		}
+	}
+
+	// Validate the arguments
+	if req.AllocID == "" {
+		handleStreamResultError(allocIDNotPresentErr, helper.Int64ToPtr(400), encoder)
+		return
+	}
+	if req.Task == "" {
+		handleStreamResultError(taskNotPresentErr, helper.Int64ToPtr(400), encoder)
+		return
+	}
+	if len(req.Cmd) == 0 {
+		handleStreamResultError(errors.New("command is not present"), helper.Int64ToPtr(400), encoder)
+	}
+
+	ar, err := a.c.getAllocRunner(req.AllocID)
+	if err != nil {
+		code := helper.Int64ToPtr(500)
+		if structs.IsErrUnknownAllocation(err) {
+			code = helper.Int64ToPtr(404)
+		}
+
+		handleStreamResultError(err, code, encoder)
+		return
+	}
+
+	capabilities, err := ar.GetTaskDriverCapabilities(req.Task)
+	if err != nil {
+		code := helper.Int64ToPtr(500)
+		if structs.IsErrUnknownAllocation(err) {
+			code = helper.Int64ToPtr(404)
+		}
+
+		handleStreamResultError(err, code, encoder)
+		return
+	}
+
+	// check node access
+	if aclObj != nil && capabilities.FSIsolation == drivers.FSIsolationNone {
+		exec := aclObj.AllowNsOp(req.QueryOptions.Namespace, acl.NamespaceCapabilityAllocNodeExec)
+		if !exec {
+			handleStreamResultError(structs.ErrPermissionDenied, nil, encoder)
+			return
+		}
+	}
+
+	allocState, err := a.c.GetAllocState(req.AllocID)
+	if err != nil {
+		code := helper.Int64ToPtr(500)
+		if structs.IsErrUnknownAllocation(err) {
+			code = helper.Int64ToPtr(404)
+		}
+
+		handleStreamResultError(err, code, encoder)
+		return
+	}
+
+	// Check that the task is there
+	taskState := allocState.TaskStates[req.Task]
+	if taskState == nil {
+		handleStreamResultError(
+			fmt.Errorf("unknown task name %q", req.Task),
+			helper.Int64ToPtr(400),
+			encoder)
+		return
+	}
+
+	if taskState.StartedAt.IsZero() {
```

Comment:
~~Hm, why do we check if the task has started? It might be useful to allow exec'ing into pending tasks that seem "frozen" due to slow artifact operations or template dependencies.~~

I figured it out! Until the driver task handle exists, there's nothing to process the exec! This check probably isn't strictly necessary given the subsequent checks for task handle, but it might be helpful.

Reply:
Yeah it's not strictly required - but this is the pattern used by other endpoints (e.g. logging and file operations) so I kept it for consistency.